### PR TITLE
feat(wpf): Add Multi-Touch support on WPF platform

### PR DESCRIPTION
**Pull request by lindexi**
GitHub Issue (If applicable): #14393

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

In the current WPF platform, if multi-touch is used, only the touch message from one finger touching the screen will be received.

## What is the new behavior?

We can receive the multi-touch events message.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features) Not need.
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable) I test pass by UnoIslandsSamplesApp.Skia.Wpf
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [x] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [x] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Reference https://github.com/unoplatform/uno/pull/15543

Cc: @MartinZikmund

Internal Issue (If applicable):

### CODE REVIEW

**Comment by lindexi** [comment]
I am currently attempting to build a test application for use on my touch device. I am encountering some issues with the UNO build, but I suspect it might be due to my local configuration. I am in the process of trying to fix it.

**Comment by lindexi** [comment]
I fixed the build issues and I test pass by UnoIslandsSamplesApp.Skia.Wpf

**Comment by MartinZikmund** [comment]
I think the problem with this approach was that the input events always eventually go to Mouse events, with StylusDevice set, so subscribing to both touch and mouse will mean all events will be triggered twice. But not completely sure, just trying to remember why I didn't go this route. Could you differentiate between touches via the StylusDevice property instead?

**Comment by lindexi** [comment]
@MartinZikmund Thank you. You're absolutely right. Whether it's touch or mouse events, they ultimately get converted into mouse events. In the WPF framework, touch events are converted into mouse events using the Promote technique. However, this conversion process results in loss of information. In the WPF framework, only the touch information from the first finger in a multi-touch scenario is converted into mouse information, while the information from other fingers is lost. Moreover, the conversion process loses touch dimensions such as width, height, and pressure sensitivity. This is why I need to separately listen to multiple events. Although listening to both Stylus and Mouse events means that I will receive two event notifications during the touch process, fortunately, I can determine in the mouse event whether the current mouse event was triggered by touch, allowing me to filter out the touch events that have already been processed.

**Comment by MartinZikmund** [comment]
@lindexi oh I see it now, filtering out events with `StylusDevice` set! That's very nice!

**Review by MartinZikmund — APPROVED** [review]

**Comment by lindexi** [comment]
Thank you @MartinZikmund

File: src/Uno.UI.Runtime.Skia.Wpf/Input/WpfCorePointerInputSource.cs [hunk 1]
```diff
@@ -230,21 +259,86 @@ private IntPtr OnWmMessage(IntPtr hwnd, int msg, IntPtr wparamOriginal, IntPtr l
...
+			};
+		}
+		else if (args is StylusEventArgs stylusEventArgs)
+		{
+			pointerId = (uint)stylusEventArgs.StylusDevice.Id;
+			position = stylusEventArgs.GetPosition(_hostControl);
+
+			properties = new()
+			{
+				IsLeftButtonPressed = true,
```
Comment (dr1rrb) [inline src/Uno.UI.Runtime.Skia.Wpf/Input/WpfCorePointerInputSource.cs:295]: For touch, this has to be `false` for release and cancel.
For pen, it has to be `true` only when `!stylusEventArgs.InAir`.

Also for the pen do we have information about the barel button and the eraser (for the `IsRightButtonPressed` and `IsEraser`)?

File: src/Uno.UI.Runtime.Skia.Wpf/Input/WpfCorePointerInputSource.cs [hunk 2]
```diff
@@ -230,21 +259,86 @@ private IntPtr OnWmMessage(IntPtr hwnd, int msg, IntPtr wparamOriginal, IntPtr l
...
+				IsPrimary = true,
+				IsInRange = !stylusEventArgs.InAir,
```
Comment (dr1rrb) [inline src/Uno.UI.Runtime.Skia.Wpf/Input/WpfCorePointerInputSource.cs:297]: This is actually always `true` except for some `Cancel` (and maybe also `Release` for touch).

File: src/Uno.UI.Runtime.Skia.Wpf/Input/WpfCorePointerInputSource.cs (hunk 1 above)
Comment (lindexi) [reply src/Uno.UI.Runtime.Skia.Wpf/Input/WpfCorePointerInputSource.cs:295]: @dr1rrb Yeah. Get the barel button by this code:

```
        private void MainWindow_StylusMove(object sender, StylusEventArgs e)
        {
            if (e.StylusDevice.StylusButtons.GetStylusButtonByGuid(StylusPointProperties.BarrelButton.Id).StylusButtonState == StylusButtonState.Down)
            {
                
            }
        }
```

Comment (lindexi) [reply src/Uno.UI.Runtime.Skia.Wpf/Input/WpfCorePointerInputSource.cs:295]: Thank you @dr1rrb  and I create the PR in https://github.com/unoplatform/uno/pull/16517 . Could you review my code?

File: src/Uno.UI.Runtime.Skia.Wpf/Input/WpfCorePointerInputSource.cs (hunk 2 above)
Comment (dr1rrb) [reply src/Uno.UI.Runtime.Skia.Wpf/Input/WpfCorePointerInputSource.cs:297]: From what I remember of my investigations, the `IsInRange` is about "do we know where the pointer currently is?", no matter if it's pressed or not.

* For mouse: does the mouse is still connected on the computer (or the USB cable as been unplugged)
* For pen: does the pen is in the detectable range (over) or touching the screen (pressed)
* For touch: does a finger is touching the screen

So when we raise a pointer event it's almost always `true` (except some `Cancel` and `Release` as mentioned). BUT you can actually keep a reference of a pointer events args (or the [`Pointer`](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.ui.xaml.input.pointerroutedeventargs.pointer?view=windows-app-sdk-1.5)) for longer than the pointer event handler scope! Properties on that args/`Pointer` will then be updated by WinUI (this is currently not supported by uno BTW), and the `IsInRange` can become `false` in those cases.

(Note: not sure about the `IsInRange` state in windowed environment, when the pointer is above another window)

**Pull request merged**